Cgen error for `fn (n &Name) str() string { ... }`

### Describe the bug

See code:

```v
module main

struct Name {
	age int = 222
}

fn (n &Name) str() string {
	return n.age.str()
}

fn main() {
	name := &Name{}
	mut mp := map[string]&Name{}
	mp['aaa'] = name

	println(mp)
}
```

### Expected Behavior

No error

### Current Behavior

```
/Users/petrmakhnev/v/v /Users/petrmakhnev/IdeaProjects/untitled -color -o /Users/petrmakhnev/IdeaProjects/untitled/bin/main
==================
/tmp/v_501/main.7482642752295985075.tmp.c:2111:53: error: passing 'main__Name' (aka 'struct main__Name') to parameter of incompatible type 'main__Name *' (aka 'struct main__Name *'); remove *
                strings__Builder_write_string(&sb, main__Name_str(**(main__Name**)DenseArray_value(&m.key_values, i)));
builder error                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/main.74: 
82642752295985075.tmp.c:1902:51: note: passing argument to parameter 'n' here
==================
VV_LOCAL_SYMBOL string main__Name_str(main__Name* n);
C error. This should never happen.
                                                  ^
                                                  ^
This is a compiler bug, please report it using `v bug file.v`./tmp/
c:2245:4: warning: expression result unused [-Wunused-value]
https://github.com/vlang/v/issues/new/choose
  (*(int*)_t1.data);
```

### Reproduction Steps

Run code above

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 d00237f.daa9034

### Environment details (OS name and version, etc.)

```
V full version: V 0.3.3 d00237f.daa9034
OS: macos, macOS, 13.0, 22A8380
Processor: 12 cpus, 64bit, little endian, Apple M2 Max

getwd: /Users/petrmakhnev/v
vexe: /Users/petrmakhnev/v/v
vexe mtime: 2023-03-14 06:34:32

vroot: OK, value: /Users/petrmakhnev/v
VMODULES: OK, value: /Users/petrmakhnev/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.37.1 (Apple Git-137.1)
Git vroot status: weekly.2023.10-50-gdaa90345
.git/config present: true

CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)
thirdparty/tcc status: thirdparty-macos-arm64 a668e5a0
```

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -703,7 +703,12 @@ fn (mut g Gen) gen_str_for_map(info ast.Map, styp string, str_fn_name string) {
 	}
 	val_styp := g.typ(val_typ)
 	mut elem_str_fn_name := val_styp.replace('*', '') + '_str'
-	if val_sym.has_method_with_generic_parent('str') {
+
+	mut receiver_is_ptr := false
+	fn_str := val_sym.find_method_with_generic_parent('str') or { ast.Fn{} }
+
+	if fn_str.name == 'str' {
+		receiver_is_ptr = fn_str.receiver_type.is_ptr()
 		match mut val_sym.info {
 			ast.Struct, ast.Interface, ast.SumType {
 				if val_sym.info.generic_types.len > 0 {
@@ -745,7 +750,7 @@ fn (mut g Gen) gen_str_for_map(info ast.Map, styp string, str_fn_name string) {
 	} else if val_sym.kind == .string {
 		tmp_str := str_intp_sq('*(${val_styp}*)DenseArray_value(&m.key_values, i)')
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${tmp_str});')
-	} else if should_use_indent_func(val_sym.kind) && !val_sym.has_method_with_generic_parent('str') {
+	} else if should_use_indent_func(val_sym.kind) && fn_str.name != 'str' {
 		ptr_str := '*'.repeat(val_typ.nr_muls())
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, indent_${elem_str_fn_name}(*${ptr_str}(${val_styp}*)DenseArray_value(&m.key_values, i), indent_count));')
 	} else if val_sym.kind in [.f32, .f64] {
@@ -759,7 +764,7 @@ fn (mut g Gen) gen_str_for_map(info ast.Map, styp string, str_fn_name string) {
 		tmp_str := str_intp_rune('${elem_str_fn_name}(*(${val_styp}*)DenseArray_value(&m.key_values, i))')
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${tmp_str});')
 	} else {
-		ptr_str := '*'.repeat(val_typ.nr_muls())
+		ptr_str := '*'.repeat(if receiver_is_ptr { val_typ.nr_muls() - 1 } else { val_typ.nr_muls() })
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${elem_str_fn_name}(*${ptr_str}(${val_styp}*)DenseArray_value(&m.key_values, i)));')
 	}
 	g.auto_str_funcs.writeln('\t\tif (i != m.key_values.len-1) {')

--- a/vlib/v/tests/ptr_str_method_test.v
+++ b/vlib/v/tests/ptr_str_method_test.v
@@ -1,0 +1,18 @@
+module main
+
+struct Name {
+	age int = 222
+}
+
+fn (n &Name) str() string {
+	return n.age.str()
+}
+
+fn test_str_method_with_ptr() {
+	name := &Name{}
+	mut mp := map[string]&Name{}
+	mp['aaa'] = name
+
+	assert mp.str() == "{'aaa': 222}"
+	assert name.str() == '222'
+}

